hamilton/plugins/dlt_extensions.py

```diff
@@ -0,0 +1,128 @@
+import dataclasses
+from typing import Any, Collection, Dict, Iterable, Literal, Optional, Sequence, Tuple, Type
+
+import dlt
+import pandas as pd
+from dlt.common.destination import Destination, TDestinationReferenceArg  # noqa: F401
+from dlt.common.schema import Schema, TColumnSchema
+
+# importing TDestinationReferenceArg fails if Destination isn't imported
+from dlt.extract.resource import DltResource
+
+from hamilton import registry
+from hamilton.io import utils
+from hamilton.io.data_adapters import DataLoader, DataSaver
+
+DATAFRAME_TYPES = [Iterable, pd.DataFrame]
+
+# TODO add types for other Dataframe libraries
+try:
+    import pyarrow as pa
+
+    DATAFRAME_TYPES.extend([pa.Table, pa.RecordBatch])
+except ModuleNotFoundError:
+    pass
+
+# convert to tuple to dynamically define type `Union[DATAFRAME_TYPES]`
+DATAFRAME_TYPES = tuple(DATAFRAME_TYPES)
+
+
+# TODO use `driver.validate_materialization`
+@dataclasses.dataclass
+class DltResourceLoader(DataLoader):
+    resource: DltResource
+
+    @classmethod
+    def name(cls) -> str:
+        return "dlt"
+
+    @classmethod
+    def applicable_types(cls) -> Collection[Type]:
+        return DATAFRAME_TYPES
+
+    def load_data(self, type_: Type) -> Tuple[pd.DataFrame, Dict[str, Any]]:
+        """Creates a pipeline and conduct `extract` and `normalize` steps.
+        Then, "load packages" are read with pandas
+        """
+        pipeline = dlt.pipeline(pipeline_name="Hamilton-DltResourceLoader")
+        pipeline.extract(self.resource)
+        normalize_info = pipeline.normalize(loader_file_format="parquet")
+
+        partition_file_paths = []
+        for package in normalize_info.load_packages:
+            load_info = package.jobs["new_jobs"][0]
+            partition_file_paths.append(load_info.file_path)
+
+        # TODO use pyarrow directly to support different dataframe libraries
+        # ref: https://arrow.apache.org/docs/python/generated/pyarrow.parquet.ParquetDataset.html#pyarrow.parquet.ParquetDataset
+        df = pd.concat([pd.read_parquet(f) for f in partition_file_paths], ignore_index=True)
```

Comment:
Parquet locally versus duckdb? Will the next `pipeline.drop()` erase it?

Reply:
So the dlt implementation for in-memory duckdb is currently bugged (they're fixing it), but I found a work around.

Using duckdb in-memory doesn't really provide additional value and only adds a dependency:
```
source -> extract (parquet) -> normalize (parquet) -> load (parquet) -> duckdb (memory) -> query db (memory) -> to pandas (memory)
```
At the end of the process, memory (duckdb, pandas) is freed and dlt pipeline is cleaned (extract, normalize, load)

My current implementation skips the duckdb steps
```
source -> extract (parquet) -> normalize (parquet) -> read parquet partitions (memory) -> pandas (memory)
```

Reply:
The related design decision is "how can I selectively reload dlt Sources". For example, only load Slack messages once then run Hamilton dataflows many times over that same data. 

This would be an ELT use case (dlt -> Hamilton) where you want to refresh each independently. It probably makes more sense to have `run_dlt.py` and `run_hamilton.py`

The current "Source" materializer with everything in-memory aims to enable ET and the user is responsible for loading the data, potentially with the Destination materializer which does TL

Reply:
can just cache the result of this function if we don't want it to run more than once?

Reply:
So yeah, the worry is that it fills up disk space. But I think we can switch this up as needed. Seems reasonable.